### DISCUSSION TITLE
OpenURL telephone number containing * -- broken in iOS v15.4.0

### DISCUSSION THREAD

**Discussion by efeder-amd** [appcelerator/titanium_mobile#13368] [Q&A] [answered]
I am using Titanium SDK 10.1.1.GA.
After I upgraded to iOS 15.4.x from 15.3.x, `Ti.Platform.openURL` stopped working for telephone numbers with a `*` in them.
I am using this to allow users of my app to place calls hiding their number using `*67` to prefix the called number.

It still works for numbers without a `*`, opening a dialog box confirming the choice to dial the number. But when the number is prefixed with `*67` it no longer works.
I am [linking](https://github.com/efeder-amd/titanium-openurl-bug-repro/blob/master/Resources/app.js) a sample app reproducing the issue. If you run on 15.3.x, both buttons work to place the call, but if you run on 15.4.x the blocked phone number no longer works.

When tracking down the code path for `openURL` I found that it [simply proxies](https://github.com/tidev/titanium_mobile/blob/b3256959bf84a6fb2e3e86554cdc0867f0d533c3/iphone/Classes/PlatformModule.m#L335) the underlying native openURL function, so this must have been an underlying Apple iOS change that broke it. 
I was researching the underlying Apple functions, and I found a couple of possibilities. I'm not really sure which one Titanium is using.

1) A method of [UIApplication](https://developer.apple.com/documentation/uikit/uiapplication/1648685-openurl/)
2) A method of [UIApplicationDelegate](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623112-application?language=objc)
3) I also saw this [open](https://developer.apple.com/documentation/uikit/uiapplication/1648685-open) method - I'm not sure what the pro's/con's would be of using this over the openURL method.

I also saw that in the [Apple URL scheme Reference](https://developer.apple.com/library/archive/featuredarticles/iPhoneURLScheme_Reference/PhoneLinks/PhoneLinks.html#//apple_ref/doc/uid/TP40007899-CH6-SW1) it specifically says that `*` and `#` characters are not supported, but this may be escaped using [stringByAddingPercentageEscapesUsingEncoding](https://developer.apple.com/documentation/foundation/nsstring/1415058-addingpercentescapes). That function is not available in titanium, but I did try encoding the `*` as `%2A` which did not work.

My issue/question is -- is there any functionality that could be added to the titanium iOS library or a way I can use hyperloop to be able to use telephone numbers prefixed with `*67`?

**Comment by caspahouzer** [comment]
Maybe something like this could work
```
import { UIApplication } from 'UIKit';
UIApplication.sharedApplication.open('tel:' + phone);
```

On the other hand, this bug must be fixed in one of the next versions. Thanks for reporting

**Reply by efeder-amd** [reply]
Thanks for the suggestion. I have to admit my understanding of how titanium objects/views map to native objects is very limited. When I try your suggestion, I get

```
[ERROR] /app.js:11
[ERROR] _UIKit.UIApplication.sharedApplication.open("tel:*672024561111");
[ERROR]                                             ^
[ERROR] TypeError: _UIKit.UIApplication.sharedApplication.open is not a function. (In '_UIKit.UIApplication.sharedApplication.open("tel:*672024561111")', '_UIKit.UIApplication.sharedApplication.open' is undefined)
[ERROR]     at  (/app.js:11:44)
[ERROR]     Foundation 0x0000000198e11ad4 _NSDescriptionWithLocaleFunc + 100
[ERROR]     CoreFoundation 0x000000019761267c D2716F2F-688B-3D51-BA8B-17E6553B209E + 42620
[ERROR]     CoreFoundation 0x000000019763147c _CFStringCreateWithFormatAndArgumentsReturningMetadata + 176
[ERROR]     TitaniumKit 0x0000000100b8d6ec TiLogMessage + 60
[ERROR]     TitaniumKit 0x0000000100bc9004 -[TiExceptionHandler reportScriptError:] + 76
[ERROR]     TitaniumKit 0x0000000100bb7028 TiBindingEventProcess + 712
[ERROR]     TitaniumKit 0x0000000100bd6ff0 -[KrollContext invoke:] + 116
[ERROR]     TitaniumKit 0x0000000100bd7060 -[KrollContext enqueue:] + 92
[ERROR]     TitaniumKit 0x0000000100bde2ac TiBindingRunLoopEnqueue + 84
[ERROR]     TitaniumKit 0x0000000100bdcad0 -[TiProxy fireEvent:withObject:propagate:reportSuccess:errorCode:message:] + 200
[ERROR]     TitaniumKit 0x0000000100bbe5d8 -[TiViewProxy fireEvent:withObject:propagate:reportSuccess:errorCode:message:] + 240
[ERROR]     BugReproCall 0x0000000100439fa8 -[TiUIButtonProxy fireEvent:withObject:propagate:reportSuccess:errorCode:message:] + 200
[ERROR]     BugReproCall 0x000000010043c258 -[TiUIButton controlAction:forEvent:] + 604
[ERROR]     UIKitCore 0x0000000199f3d1fc E6E89BB5-B586-3725-BF8F-4C85DCD443BD + 4997628
[ERROR]     UIKitCore 0x000000019a067bb4 E6E89BB5-B586-3725-BF8F-4C85DCD443BD + 6220724
[ERROR]     UIKitCore 0x0000000199de4bc4 E6E89BB5-B586-3725-BF8F-4C85DCD443BD + 3587012
[ERROR]     UIKitCore 0x0000000199e818d4 E6E89BB5-B586-3725-BF8F-4C85DCD443BD + 4229332
[ERROR]     UIKitCore 0x000000019a112c30 E6E89BB5-B586-3725-BF8F-4C85DCD443BD + 6921264
[ERROR]     UIKitCore 0x0000000199be9e60 E6E89BB5-B586-3725-BF8F-4C85DCD443BD + 1511008
[ERROR]     UIKitCore 0x0000000199c1b100 E6E89BB5-B586-3725-BF8F-4C85DCD443BD + 1712384
```

**Reply by caspahouzer** [reply]
Sorry, I've got no more idea. But Maybe someone of the https://slack.tidev.io community can help. I will post this discussion there

**Comment by hansemannn** [comment]
Hi there! Do you have an example phone number to test with? And I assume you did test on a physical device? The correct Hyperloop code for the above would be this one btw:
```js
UIApplication.sharedApplication.openURL('tel:' + phone);
```
But lets fix it natively!

*UPDATE*: It seems like a known iOS issue, see [this thread](https://developer.apple.com/forums/thread/701865) for details.

**Reply by efeder-amd** [reply]
Ahh thank you for pointing me to the iOS bug thread. It seems at this point there is no native fix :(

**Reply by caspahouzer** [reply]
What is a use-case für numbers with * ?

**Reply by m1ga** [reply]
`to place calls hiding their number` apparently you can add this before and it will turn off showing the number.

**Comment by joshualambert** [comment]
@efeder-amd Just dropping in here for a second to say **_EXCELLENT_** work on your bug report here. This is one of the better ones I've seen on Github.

**Comment by efeder-amd** [comment, marked as answer]
Latest update -- I've opened both a [Technical Support Incident](https://developer.apple.com/support/technical/) and submitted the bug via [Feedback Assistant](https://feedbackassistant.apple.com/).
From the TSI I got the response:
"Our engineers have reviewed your request and have determined that you are experiencing a known issue for which there is no known workaround at this time."
Knowing that there was no native option, my team ended up using Twilio to route the call and spoof the callback number (which is an added cost to us versus the user's phone company doing that, but a necessary app requirement).
